### PR TITLE
Add mirror support for input files

### PR DESCRIPTION
**Pull request by douglasjacobsen**
This merge adds mirror support to ramble.

Mirrors can be used to replicate input files for workloads. Currently, `mirror create` requires an active workspace. Workloads that are used within the workspace have their input files replicated into the mirror when `mirror create` is called.

Additionally, this merge adds support for an input cache, where input files are downloaded before being used within a workspace. This minimized the number of times a file needs to be downloaded when being used in many workspaces.

### CODE REVIEW

**Review by rfbgo — DISMISSED** [review]
LGTM

Honestly this PR is too big for me to be super confident in my review, but it's a great piece of functionality

File: lib/ramble/ramble/cmd/mirror.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+    # Destroy
+    destroy_parser = sp.add_parser('destroy', help=mirror_destroy.__doc__)
+
+    destroy_target = destroy_parser.add_mutually_exclusive_group(required=True)
+    destroy_target.add_argument('-m', '--mirror-name',
+                                metavar='mirror_name',
+                                type=str,
+                                help="find mirror to destroy by name")
+    destroy_target.add_argument('--mirror-url',
```
[the remark this answers is not in the document]
Comment (douglasjacobsen) [reply lib/ramble/ramble/cmd/mirror.py]: Done!

File: lib/ramble/ramble/cmd/mirror.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    if args.push:
+        changes_made = changes_made or push_url != url
+        push_url = url
+    else:
+        changes_made = changes_made or push_url != url
+        fetch_url, push_url = url, url
+
+    items = [
+        (
+            (n, u)
```
[the remark this answers is not in the document]
Comment (douglasjacobsen) [reply lib/ramble/ramble/cmd/mirror.py]: `-n` is already taken as implying no checksums.

I updated these though.

File: lib/ramble/ramble/caches.py [hunk 3]
```diff
@@ -34,3 +36,58 @@ def _misc_cache():
...
+    def store(self, fetcher, relative_dest):
+        """Fetch and relocate the fetcher's target into our mirror cache."""
+
+        # Note this will archive package sources even if they would not
+        # normally be cached (e.g. the current tip of an hg/git branch)
+        dst = os.path.join(self.root, relative_dest)
+        mkdirp(os.path.dirname(dst))
+        fetcher.archive(dst)
+
+    def symlink(self, mirror_ref):
```
[the remark this answers is not in the document]
Comment (douglasjacobsen) [reply lib/ramble/ramble/caches.py:72]: This is copied from Spack, and used in other classes. I'm going to leave it for now, just to (hopefully) make it easier to update to newer versions of spack in the future.

File: lib/ramble/ramble/mirror.py [hunk 4]
```diff
@@ -0,0 +1,608 @@
...
+
+    def to_dict(self):
+        if self._push_url is None:
+            return syaml_dict([
+                ('fetch', self._fetch_url),
+                ('push', self._fetch_url)])
+        else:
+            return syaml_dict([
+                ('fetch', self._fetch_url),
+                ('push', self._push_url)])
```
[the remark this answers is not in the document]
Comment (douglasjacobsen) [reply lib/ramble/ramble/mirror.py]: I like the last of the ones you suggested. I went ahead and updated it to that.

File: lib/ramble/ramble/mirror.py [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+    def __eq__(self, other):
+        return self._mirrors == other._mirrors
+
+    def to_json(self, stream=None):
+        return spack.util.spack_json.dump(self.to_dict(True), stream)
+
+    def to_yaml(self, stream=None):
+        return spack.util.spack_yaml.dump(self.to_dict(True), stream)
+
+    # TODO: this isn't called anywhere
```
[the remark this answers is not in the document]
Comment (douglasjacobsen) [reply lib/ramble/ramble/mirror.py]: Removed! (And some tests are removed, so I added some other tests to help increase coverage)

**Comment by douglasjacobsen** [comment]
:) A lot of it is just migrating functionality from Spack proper.

**Pull request merged**